### PR TITLE
Add a new module to logout inactive session after a configured delay

### PR DESCRIPTION
**Pull request by Cedric-Pigeon**
The goal of this PR is to set a validity on a user session.

The validity delay is set in a system parameter and its default value is 2 hours.
If the user do nothing during this delay, his session is automatically logged out and he must re-login.

### CODE REVIEW

File: inactive_session_timeout/models/res_users.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+from os.path import getmtime
+from time import time
+
+DELAY_KEY = 'inactive_session_time_out_delay'
+IGNORED_PATH_KEY = 'inactive_session_time_out_ignored_url'
+
+
+class res_users(models.Model):
+    _inherit = 'res.users'
+    _REMOTE_USER_ATTRIBUTE = 'HTTP_REMOTE_USER'
```
Comment (lmignon) [inline inactive_session_timeout/models/res_users.py]: Not used

Comment (Cedric-Pigeon) [inline inactive_session_timeout/models/res_users.py]: Removed...

File: inactive_session_timeout/README.rst [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+This module was written to be able to kill(logout) all inactive sessions since
+a given delay. On each request the server checks if the session is yet valid
+regarding the expiration delay. If not a clean logout is operated.
+
+Configuration
+=============
+
+Two system parameters are available:
+
+* inactive_session_time_out_delay: validity of a session in secondes (default = 2 Hours)
```
Comment (hbrunn) [inline inactive_session_timeout/README.rst]: seconds (without the second 'e')

File: inactive_session_timeout/models/res_users.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+class res_users(models.Model):
+    _inherit = 'res.users'
+
+    def _check_session_validity(self, db, uid, passwd):
+        if not request:
+            return
+        session = request.session
+        session_store = root.session_store
+        registry = self.pool
+        if InactiveSessionTimeoutInstalled._name not in registry:
```
Comment (hbrunn) [inline inactive_session_timeout/models/res_users.py]: Why do you need this check if this is not a server wide module?

File: inactive_session_timeout/models/res_users.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            # module not installed in database,
+            # continue usual behavior
+            return
+        param_model = registry['ir.config_parameter']
+        cr = self.pool.cursor()
+        try:
+            delay = int(param_model.get_param(cr, uid, DELAY_KEY, 7200))
+            urls = param_model.get_param(cr, uid, IGNORED_PATH_KEY,
+                                         '').split(',')
```
Comment (hbrunn) [inline inactive_session_timeout/models/res_users.py]: this causes 2 extra db requests per http requests, shouldn't you better cache this values?

**Comment by hbrunn** [comment]
On a general note, if I needed this functionality I'd look what configuration parameters werkzeug has here. Did you do that? I honstely don't know if werkzeug provides anything, but I have the feeling this would be the right place to look

File: inactive_session_timeout/README.rst (hunk 2 above)
Comment (Cedric-Pigeon) [inline inactive_session_timeout/README.rst]: Fixed

File: inactive_session_timeout/models/res_users.py (hunk 3 above)
Comment (Cedric-Pigeon) [inline inactive_session_timeout/models/res_users.py]: This was some remaining code due to a previous try. I removed it.

File: inactive_session_timeout/models/res_users.py (hunk 4 above)
Comment (Cedric-Pigeon) [inline inactive_session_timeout/models/res_users.py]: Done.

**Comment by Cedric-Pigeon** [comment]
@hbrunn I tried to act earlier in the server on the http request but I didn't find any single entry point for all odoo requests. This is way I override the check method of res.user.

File: inactive_session_timeout/README.rst [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+regarding the expiration delay. If not a clean logout is operated.
+
+Configuration
+=============
+
+Two system parameters are available:
+
+* inactive_session_time_out_delay: validity of a session in seconds (default = 2 Hours)
+* inactive_session_time_out_ignored_url: technical urls where the check does not occur 
+
```
Comment (adrienpeiffer) [inline inactive_session_timeout/README.rst]: Can you add bugt tracker section ?

File: inactive_session_timeout/__openerp__.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+#     If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+{
+    'name': "Inactive Sessions Timeout",
+
+    'summary': """
+        This module disable all inactive sessions since a given delay""",
+
+    'author': "ACSONE SA/NV",
```
Comment (adrienpeiffer) [inline inactive_session_timeout/__openerp__.py]: Add OCA as author like `ACSONE SA/NV,Odoo Community Association (OCA)`

File: inactive_session_timeout/README.rst [hunk 7]
```diff
@@ -0,0 +1,35 @@
+Inactive Sessions Timeout
```
Comment (lmignon) [inline inactive_session_timeout/README.rst]: can you add the AGPL-3 badge

**Comment by adrienpeiffer** [comment]
I restarted runbot build

File: inactive_session_timeout/models/ir_config_parameter.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+from openerp import models, api
+
+from openerp import tools
+from openerp import SUPERUSER_ID
+
+DELAY_KEY = 'inactive_session_time_out_delay'
+IGNORED_PATH_KEY = 'inactive_session_time_out_ignored_url'
+
+
+class ir_config_parameter(models.Model):
```
Comment (pedrobaeza) [inline inactive_session_timeout/models/ir_config_parameter.py]: Use CamelCase for class name

File: inactive_session_timeout/models/res_users.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+from openerp.http import root
+from openerp.http import request
+
+from os import utime
+from os.path import getmtime
+from time import time
+
+
+class res_users(models.Model):
```
Comment (pedrobaeza) [inline inactive_session_timeout/models/res_users.py]: CamelCase

**Comment by pedrobaeza** [comment]
Interesting module. Little remarks, but in general looks good

File: inactive_session_timeout/models/ir_config_parameter.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+#     GNU Affero General Public License
+#     along with inactive_session_timeout.
+#     If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+from openerp import models, api
+
+from openerp import tools
+from openerp import SUPERUSER_ID
+
```
Comment (alejandrosantana) [inline inactive_session_timeout/models/ir_config_parameter.py:26]: Wouldn't it be OK in just one line?:
`from openerp import api, models, SUPERUSER_ID, tools`

**Comment by alejandrosantana** [comment]
Wouldn't it be enough overriding `openerp/http.py` and changing the hardcoded value 60*60*24*7 with the value of your parameter? Or also helpful?
https://github.com/odoo/odoo/blob/8.0/openerp/http.py#L1164

``` python

def session_gc(session_store):
    if random.random() < 0.001:
        # we keep session one week
        last_week = time.time() - 60*60*24*7
        for fname in os.listdir(session_store.path):
            path = os.path.join(session_store.path, fname)
            try:
                if os.path.getmtime(path) < last_week:
                    os.unlink(path)
            except OSError:
                pass
```

**Comment by Cedric-Pigeon** [comment]
@lmignon Done. Thanks for reviewing.
@adrienpeiffer Done. Thanks for reviewing.
@pedrobaeza Done. Thanks for reviewing.
@alejandrosantana Done. Thanks for reviewing.

**Comment by nhomar** [comment]
:+1:

**Comment by pedrobaeza** [comment]
:+1:

File: inactive_session_timeout/README.rst [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+Two system parameters are available:
+
+* inactive_session_time_out_delay: validity of a session in seconds (default = 2 Hours)
+* inactive_session_time_out_ignored_url: technical urls where the check does not occur 
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/{project_repo}/issues>`_.
```
Comment (laetitia-gangloff) [inline inactive_session_timeout/README.rst]: replace value in {}

**Comment by Cedric-Pigeon** [comment]
@laetitia-gangloff Done, thanks.

**Comment by laetitia-gangloff** [comment]
:+1:

**Comment by alejandrosantana** [comment]
@Cedric-Pigeon Nice module. I am testing it functionally, and looks fine.

:+1:

**Pull request merged**